### PR TITLE
Add support for recursive references in object types

### PR DESCRIPTION
**Pull request by simonsinclair**
Resolves https://github.com/fastify/fast-json-stringify/issues/365.

#### Notes
- This probably needs more tests.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: test/recursion.test.js [hunk 1]
```diff
@@ -135,3 +135,46 @@ test('use proper serialize function', t => {
...
+            $ref: '#/$defs/parent_category'
+          }
+        }
+      }
+    }
+  }
+
+  const stringify = build(schema)
+  const data = {
+    category: {
```
Comment (climba03003) [inline test/recursion.test.js:168]: Can you add an unit test for
```
const parant = {
  parant: parant
}
const data = {
  category: {
    parant: parant
  }
}
```

I thiink most of the recursive reference support library will face this unless object reference looping.

File: test/recursion.test.js [hunk 2]
```diff
@@ -135,3 +135,46 @@ test('use proper serialize function', t => {
...
+      category: {
+        type: 'object',
+        properties: {
+          parent: {
+            $ref: '#/$defs/parent_category'
+          }
+        }
+      }
+    },
+    $defs: {
```
Comment (Eomm) [inline test/recursion.test.js:154]: please use JSON schema draft 07 syntax

**Review by mcollina — APPROVED** [review]
lgtm

File: test/recursion.test.js (hunk 1 above)
Comment (simonsinclair) [reply test/recursion.test.js:168]: > Can you add an unit test for
> 
> ```
> const parant = {
>   parant: parant
> }
> const data = {
>   category: {
>     parant: parant
>   }
> }
> ```

Is this a valid use-case? Attempting to reference "`parant`" on L2 before it's defined will result in an error.

**Review by L2jLiga — APPROVED** [review]
LGTM

**Comment by mcollina** [comment]
is this ready to land?

**Comment by L2jLiga** [comment]
> is this ready to land?

From my end I see that tests in CI passed and we're step closer to give recursion support, in future we can think about how to have only one map, but for now it's fine for me

**Comment by simonsinclair** [comment]
I had a test in mind for multiple types, which is currently failing, i.e.:

```js
parentCategory: {
  type: ['object', 'null'], // 👈
  properties: {
    parent: {
      $ref: '#/definitions/parentCategory'
    }
  }
}
```

I'm looking into this now, but happy build on this in another PR if you prefer.

**Comment by L2jLiga** [comment]
I have no strong opinion on that, before we had crash, now we can work with this data, even if we have some bugs or mistakes they shouldn't affect functionality that was working and affect only recursion support

File: test/recursion.test.js (hunk 1 above)
Comment (climba03003) [reply test/recursion.test.js:168]: My mistake on the example, it should works this time.
```
const parant = {}
const data = {
  category: {  parant }
}
parant.parant = parant
```

**Pull request merged**